clang/include/clang/Basic/DiagnosticSemaKinds.td

```diff
@@ -7306,6 +7306,9 @@ let CategoryName = "Lambda Issue" in {
     "here">;
   def note_var_explicitly_captured_here : Note<"variable %0 is"
     "%select{| explicitly}1 captured here">;
+  def err_implicit_this_capture : Error<
+    "implicit capture of 'this' with a capture default of '=' is not allowed "
+    "for kernel functions in SYCL 1.2.1">;
```

Comment:
I'm not sure that we need to mention `=` in this diagnostic at all, because AFAIK SYCL kernel lambdas can be defined only with `=`.

Reply:
Do we prohibit & capture?  I realize that we can't capture references (right?!), but & capture of 'this' as a slightly different meaning.

Reply:
We diagnose captured references (I mean the fields of kernel object that have reference type), but I don't remember diagnostic for the whole `&` lambda capture. Isn't it covered by captured references diagnostic? Should we diagnose it separately?

Reply:
THAT covers it for the most part (thought it might be confusing since adding a line to a lambda will cause it to actually capture by reference), but capturing 'this' is a confusing story.  Capturing 'this' by reference is exactly the same as not capturing by this, since the pointer cannot change.

We likely want to change the error message to say 'implicit capture of this', which is what the rest of the patch implements.

Reply:
Is there a reason only implicit capture of `this` is not allowed? I would assume whatever issues arise from capturing `this` would exist for implicit and explicit capture?

Reply:
Good question. I somehow was under impression that kernel lambdas can use only implicit capture, however I've opened the spec where it describes how to write kernel lambda and it says (pt. 4.8.9.2):
>The kernel lambda must use copy for all of its captures (i.e. [=]).

For me this statement doesn't prohibit copy for explicit captures.

Reply:
Yes, there is no reason to forbid explicit captures.